### PR TITLE
fix(inventory): Fix inventory source caching

### PR DESCRIPTION
**Pull request by 0xFelix**
**What this PR does / why we need it**:

Fix inventory source caching by separating the fetching of objects and
populating the inventory. This way objects can be fetched from the K8S
API or from a configured cached and the cache related parameters on the
plugin now actually work.

To make this possible use dicts instead of ResourceFields where possible 
to allow easier serialization/deserialization of objects fetched from
the K8S API.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

Merge after #117 

**Release note**:

```release-note
Caching of inventory sources was fixed and is now working.
```

### CODE REVIEW

File: plugins/inventory/kubevirt.py [hunk 1]
```diff
@@ -375,61 +387,49 @@ def parse(self, inventory: Any, loader: Any, path: str, cache: bool = True) -> N
...
-            opts.base_domain = self.get_cluster_domain(client)
-        for namespace in namespaces:
-            self.populate_inventory_from_namespace(client, name, namespace, opts)
+
+        results = {}
+        if attempt_to_read_cache:
+            try:
+                results = self._cache[cache_key]
+            except KeyError:
+                cache_needs_update = True
```
Comment (jcanocan) [inline plugins/inventory/kubevirt.py:415]: I'm not 100% sure, but IMHO, using exceptions for checking if an element is present or not in a data structure is not a good practice. What about:
```python
if attempt_to_read_cache:
     if cache_key in self._cache:
         results = self._cache[cache_key]
     else:
         cache_needs_update = True
```
Using exceptions as flow control, generally speaking, it is not a good idea.

Comment (jcanocan) [reply plugins/inventory/kubevirt.py:415]: results = self._cache.get(cache_key, {})
And checking if the `result` var is `{}` and updating the cache in that case may also work.

**Review by jcanocan — COMMENTED** [review]
Nice job! Just a couple of comments.

File: plugins/inventory/kubevirt.py [hunk 2]
```diff
@@ -375,61 +387,49 @@ def parse(self, inventory: Any, loader: Any, path: str, cache: bool = True) -> N
...
 
-    def fetch_objects(self, config_data: Dict) -> None:
-        """
-        fetch_objects populates the inventory with the specified parameters.
-        """
-        if not config_data or not isinstance(config_data, dict):
-            config_data = {}
+        config_data = self._read_config_data(path)
+        cache_key = self.get_cache_key(path)
+        user_cache_setting = self.get_option("cache")
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply plugins/inventory/kubevirt.py:403]: If not defined this returns `None` which evaluates to the default of `False`.

File: plugins/inventory/kubevirt.py [hunk 3]
```diff
@@ -375,61 +387,49 @@ def parse(self, inventory: Any, loader: Any, path: str, cache: bool = True) -> N
...
+        attempt_to_read_cache = user_cache_setting and cache
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply plugins/inventory/kubevirt.py:404]: No, the default is to not cache it.

File: tests/unit/plugins/inventory/test_kubevirt_parse.py [hunk 4]
```diff
@@ -13,39 +13,145 @@
...
+
+
+@pytest.mark.parametrize(
+    "cache_parse,cache_option,cache_data,expected",
+    [
+        (True, True, {"test-key": {"something": "something"}}, True),
+        (False, True, {"test-key": {"something": "something"}}, False),
+        (True, False, {"test-key": {"something": "something"}}, False),
+        (False, False, {"test-key": {"something": "something"}}, False),
+        (True, True, {"test-key2": {"something": "something"}}, False),
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply tests/unit/plugins/inventory/test_kubevirt_parse.py:120]: Added

File: plugins/inventory/kubevirt.py (hunk 1 above)
Comment (0xFelix) [reply plugins/inventory/kubevirt.py:415]: `self._cache` is not a plain dict and therefore might behave slightly different. This was taken from the Ansible documentation.

https://docs.ansible.com/ansible/9/dev_guide/developing_inventory.html#inventory-cache

**Review by lyarwood — COMMENTED** [review]
/approve
/hold

This looks correct after skimming over the main changes.

@jcanocan can I leave you to unhold and lgtm once you're happy? 

FWIW we really should make you an approver of this project as you've built up much more context and awareness than me. I feel like I'm not keeping up with the great work you're both doing tbh.

File: plugins/inventory/kubevirt.py (hunk 1 above)
Comment (jcanocan) [reply plugins/inventory/kubevirt.py:415]: Ah, I see. Sounds good them.

**Comment by jcanocan** [comment]
> /approve /hold
> 
> This looks correct after skimming over the main changes.
> 
> @jcanocan can I leave you to unhold and lgtm once you're happy?
> 
> FWIW we really should make you an approver of this project as you've built up much more context and awareness than me. I feel like I'm not keeping up with the great work you're both doing tbh.

Thanks a lot for the proposal! I will do my very best :)

**Comment by jcanocan** [comment]
Nice job!
I leave you @0xFelix unhold.
/lgtm

**Pull request merged**